docs/getting-started/roles/sqlalchemy/basic_roles.rst

```diff
@@ -0,0 +1,270 @@
+
+
+
+Add an endpoint that needs authorization
+----------------------------------------
+
+Create or choose an existing endpoint that will need authorization to access.
+In our sample app we've created 2 endpoints that have different authorization
+requirements: one to view repositories and another to view billing
+information.
+
+Add policy checks to your code to control access to the protected endpoints.
+
+
+.. literalinclude:: /examples/roles/sqlalchemy/oso-sqlalchemy-roles-guide/app/routes.py
+    :class: copybutton
+    :language: python
+    :lines: 1-2, 5-6, 13-27
+    :caption: :fab:`python` routes.py
+
+Our example uses :py:func:`flask_oso.FlaskOso.authorize` to complete the
+policy check. If you're not using Flask, check out our general-purpose
+:doc:`Python package </using/libraries/python/index>`.
+
+2. Add roles
+============
+
+Create the OrganizationRole class using the role mixin
+------------------------------------------------------
+
+The oso SQLAlchemy library provides a method to generate a mixin which
+creates a role model. Create the mixin by passing in the base, user, and
+organization models, as well as the role names. Then create a role model that
+extends it.
+
+
+.. literalinclude:: /examples/roles/sqlalchemy/oso-sqlalchemy-roles-guide/app/models.py
+    :class: copybutton
+    :language: python
+    :lines: 6, 50-58
+    :caption: :fab:`python` models.py
+
+
+
+Specify role permissions
+------------------------
+
+To give the roles permissions, write an oso policy.
+
+Since we already called :py:func:`sqlalchemy_oso.roles.enable_roles` in our ``init_oso()`` method,
+you can write Polar ``role_allow`` rules over ``OrganizationRoles``.
+
+.. literalinclude:: /examples/roles/sqlalchemy/oso-sqlalchemy-roles-guide/app/authorization.polar
+    :class: copybutton
+    :language: polar
+    :lines: 7-14
+    :caption: :fa:`oso` authorization.polar
+
+
+You can also specify a :ref:`hierarchical role ordering <role-hierarchies>` with ``organization_role_order`` rules.
+
+.. literalinclude:: /examples/roles/sqlalchemy/oso-sqlalchemy-roles-guide/app/authorization.polar
+    :class: copybutton
+    :language: polar
+    :lines: 24-26
+    :caption: :fa:`oso` authorization.polar
+
+
+For more details on the roles base policy, see :doc:`/getting-started/builtin-roles/index`.
+
+
+Create an endpoint for assigning roles
+--------------------------------------
+
+Until you assign users to roles, they'll receive a ``403 FORBIDDEN`` response if they try to
+access either protected endpoint.
+
+Add a new endpoint to your application that users can hit to assign roles. To
+control who can assign roles, add another call to
+:py:func:`flask_oso.FlaskOso.authorize`.
+
+Use the oso role API to create role assignments with
+:py:meth:`sqlalchemy_oso.roles.add_user_role` and
+:py:meth:`sqlalchemy_oso.roles.reassign_user_role`.
+
+.. literalinclude:: /examples/roles/sqlalchemy/oso-sqlalchemy-roles-guide/app/routes.py
```

Comment:
Why is an exception raised in this code? I don't see that behavior documented.

Reply:
A `session.commit` is required somewhere to save this role assignment, right? Does that happen in the after request handler in this example app?

Reply:
An exception is raised because of the unique constraints. Adding comments. Commit is done in the method (see the `commit` flag)